provider/pkg/schema/extension_resource.go

```diff
@@ -0,0 +1,272 @@
+	}
+	if !exists {
+		return nil, nil, false, nil
+	}
+
+	// Fill in the write-only properties from the old state as they won't included when reading.
+	if len(extensionResource.WriteOnly) > 0 {
+		missingProps := make([]string, 0, len(extensionResource.WriteOnly))
+		for _, writeOnlyProp := range extensionResource.WriteOnly {
+			if _, ok := resourceState[writeOnlyProp]; !ok {
+				oldValue, ok := extensionResource.Properties[writeOnlyProp]
+				missingProps = append(missingProps, writeOnlyProp)
+				if ok {
+					resourceState[writeOnlyProp] = oldValue
+				}
+			}
+		}
+		r.host.Log(ctx, diag.Warning, urn, fmt.Sprintf("Can't refresh write-only properties: %s", strings.Join(missingProps, ", ")))
+	}
+
+	newOutputs := resource.NewPropertyMapFromMap(resourceState)
+	oldProperties := resource.NewPropertyMapFromMap(extensionResource.Properties)
+	newProperties := createNewInputs(oldProperties, oldOutputsMap, newOutputs)
+	extensionResource.Properties = resourcex.Decode(newProperties)
+	newInputs := oldInputs.Copy()
+	newInputs[resource.PropertyKey("properties")] = resource.PropertyValue{V: newProperties}
+	return extensionResource.ToOutputs(resourceState), newInputs, true, nil
+}
+
+func (r *extensionResource) Update(ctx context.Context, urn resource.URN, id string, inputs resource.PropertyMap, oldInputs resource.PropertyMap) (resource.PropertyMap, error) {
+	var oldExtensionResource ExtensionResource
+	_, err := resourcex.Unmarshal(&oldExtensionResource, oldInputs, resourcex.UnmarshalOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal old inputs: %w", err)
+	}
+
+	var newExtensionResource ExtensionResource
+	_, err = resourcex.Unmarshal(&newExtensionResource, inputs, resourcex.UnmarshalOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal new inputs: %w", err)
+	}
+
+	if oldExtensionResource.Type != newExtensionResource.Type {
+		return nil, fmt.Errorf("changing the type of an extension resource is not supported")
+	}
+
+	jsonDiffPatch, err := jsondiff.Compare(oldExtensionResource.Properties, newExtensionResource.Properties)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compare properties: %w", err)
+	}
+
+	// Write-only properties can't even be read internally within the CloudControl service so they must be included in
+	// patch requests as adds to ensure the updated model validates.
+	for _, writeOnlyPropName := range newExtensionResource.WriteOnly {
+		newValue, ok := newExtensionResource.Properties[writeOnlyPropName]
+		if !ok {
+			continue
+		}
+		hasPatch := false
+		for _, op := range jsonDiffPatch {
+			if op.Path == writeOnlyPropName {
+				hasPatch = true
+				break
+			}
+		}
+		if !hasPatch {
+			jsonDiffPatch = append(jsonDiffPatch, jsondiff.Operation{
+				Type:  "add",
+				Path:  writeOnlyPropName,
+				Value: newValue,
+			})
+		}
+	}
+
+	jsonPatch := make([]jsonpatch.JsonPatchOperation, 0, len(jsonDiffPatch))
+	for _, op := range jsonDiffPatch {
```

Comment:
How is the diff calculation specific to the extension resources? Don't we take exactly the same steps for "normal" resources? Why can't the provider do that and then pass the diff down to the resource?

Reply:
No, we can't use the same diff & patch implementation because that requires us to have metadata which includes the full schema for the resource and its types - which would not be ergonomic for a user to provide.

Reply:
Interestingly ... this might actually be useful to port back to the general resource too as it looks like a better overall implementation, but it's not required right now as it seems to be functioning okay.